keps/sig-api-machinery/customresources-pruning-and-defaulting.md

```diff
@@ -0,0 +1,486 @@
+
+Note that we lose expressiveness of the existing `format` strings: we either apply the format to `"json"` or any other pre-defined format. I.e. we cannot express that pruning should be disabled, but if it is an integer, it should be an int32.
+
+**Question:** this feels like a reasonable loss of expressivity. Do we accept that?
+
+**Raw JSON Option 2:** add an extension property, e.g.
+
+```json
+{"properties": {"x": {"x-no-pruning": true}}}
+```
+
+This would not lead to a loss in expressivity. It is formulated negatively intentionally to have `false` as the default with `omitempty`.
+
+As OpenAPI extensions are prefixed with `x-kubernetes` elsewhere, we probably should name it `x-kubernetes-no-pruning: true`.
+
+**We propose to follow option with the `x-kubernetes` prefix** because it is more explicit, does not reduce expressivity and does not mangle with the already very vague `format` field definition.
+
+### Nested x-no-pruning
+
+**Question:** should we support nested `x-no-pruning`, i.e. disabling pruning for a sub-object, but re-enable it something deep inside of it? E.g.
+
+```json
+{
+    "properties": {
+        "x": {
+            "x-no-pruning": true,
+            "properties": {
+		        "y": { 
+                    "x-kubernetes-no-pruning": false,
+                    "properties": { "z": {} }
+                }
+            }
+        }
+    }
+}
+```
+
+If we do, the object 
+
+```json
+{
+    "a": 1,
+    "x": {
+        "b": 2,
+        "y": {
+            "c": 3,
+            "z": 42
+        }
+    }
+}
+```
+
+is pruned to  `{"x":{"b": 2, "y":{"z": 42}}}`.
+
+## Related Topics
+
+### Ratcheting (slightly related)
+
+In https://github.com/kubernetes/kubernetes/pull/64907 a ratcheting mechanism during validation is introduced: if a specific proposition during validation is not fulfilled in an object stored in etcd, it is not required either during updates of that object.
+
+Ratcheting does not work well with arbitrary propositional OpenAPI validation schemata because propositions in negative positions (under `not` or `oneOf`) that are weakened via ratcheting will make the whole validation stricter instead of weaker (compare https://github.com/kubernetes/kubernetes/pull/64907#issuecomment-396526377).
+
+We have multiple options to restrict the use of ratcheting (which could be the property of certain OpenAPI validation schemata constructs):
+1. Allow ratcheting only in the skeleton schema. Everything in the skeleton schema is positive.
+2. Disallow ratcheting under `not` and under `oneOf`, but allow it everywhere else.
+3. Disallow ratcheting in negative positions. This implies that it is disallowed under `oneOf`, but is allowed e.g. under `not: {not: ….}`.
+The first is the most strict one, the last is the most relaxed one. On the other hand, 1 is the simplest to reason about.
```

Comment:
**Nikhita Raghunath:** If I were using this, I would prefer 1 because it is easy to reason about and IMO if the author is changing the schema, they are aware that they might break their users...I think 1 should be sufficient.

Would be nice to factor in @liggitt 's views on this too.

Reply:
there was more discussion on the ratcheting issue/PR at https://github.com/kubernetes/kubernetes/issues/64841#issuecomment-405431399, specifically around:
* what cases we allow tightening validating for (current plan is only for problems that result in fatally flawed objects)
* when ratcheting validation can be promoted to standard validation (only once we can verify there is no persisted data that would fail the tighter check)

Reply:
Ok, looks like ratcheting for native type is up for debate. I will leave this section here. In the non-goal we write that implementing ratcheting for CRDs is out of scope here. So this is merely a comment to be in-line with ratcheting if it is ever agreed on.